usr/lib/python3/dist-packages/stdisplay/stdisplay.py

```diff
@@ -0,0 +1,275 @@
+##
+## SPDX-License-Identifier: AGPL-3.0-or-later
+
+"""
+Sanitize text to be safely printed to the terminal.
+"""
+
+from curses import setupterm, tigetnum, error as curses_error
+from os import environ
+from re import compile as re_compile, sub as re_sub
+from typing import Optional
+
+
+def get_sgr_support() -> int:
+    """Returns number of supported SGR codes.
+
+    Number of supported colors is limited to a number, there is not definition
+    of which SGR queries are supported without trying to parse the terminfo
+    database.
+
+    To disable SGR support:
+
+    *   Set a non-empty value to the environment variable NO_COLOR; or
+    *   Set the environment variable TERM to a terminal that doesn't support
+        SGR, such as 'dumb'.
+
+    Returns
+    -------
+    int
+        Number of supported SGR codes.
+
+    Notes
+    -----
+    Terminfo database is outdated, it may support 24-bit mode and not advertise
+    it. Some terminals can convert higher bit modes to lower ones, but there is
+    no easy way to know if it is supported or not.
+    https://github.com/termstandard/colors?tab=readme-ov-file#querying-the-terminal
+
+    It is possible to query the terminfo database from the shell:
+
+    $ tput -T "$TERM" colors
+
+    $ for term in $(toe -as | awk '{print $2}'); do
+        printf '%d %s\n' "$(tput -T "$term" colors || printf '-1')" "$term"
+      done | sort -n
+
+    Examples
+    --------
+    Get SGR code count supported by your terminal:
+    >>> get_sgr_support()
+    8
+
+    Get SGR code count for any terminal:
+    >>> import os
+    >>> os.environ["NO_COLOR"] = ""
+    >>> os.environ["COLORTERM"] = ""
+    >>> os.environ["TERM"] = "xterm-16color"
+    >>> get_sgr_support()
+    16
+
+    Disable SGR codes:
+    >>> import os
+    >>> os.environ["NO_COLOR"] = "1"
+    >>> get_sgr_support()
+    -1
+    """
+    if environ.get("NO_COLOR"):
+        return -1
+    if environ.get("COLORTERM") in ("truecolor", "24bit"):
+        return 2**24
+    try:
+        setupterm()
+        return tigetnum("colors")
+    except curses_error:
+        return -2
+
+
+def exclude_pattern(original_pattern: str, negate_pattern: list[str]) -> str:
+    """Exclude matching next expression if provided expression matches.
+
+    Parameters
+    ----------
+    original_pattern : str
+        Pattern which contains unwanted matches.
+    negate_pattern : list[str]
+        Pattern to exclude from the next match.
+
+    Returns
+    -------
+    str
+        Regular expression with negative lookahead.
+
+    Examples
+    --------
+    Patterns for 3-bit and 4-bit SGR is simple:
+    >>> exclude_pattern(r"(0*(30|31))", ["0*31"])
+    '(?!(?:0*31))(0*(30|31))'
+
+    When excluding 8-bit and 24-bit SGR, a wild matches can be used:
+    >>> exclude_pattern(r"(0*38(:|;)0*5(:|;)0*[0-9])", ["0*38;0*5(:|;)[0-9]+"])
+    '(?!(?:0*38;0*5(:|;)[0-9]+))(0*38(:|;)0*5(:|;)0*[0-9])'
+    """
```

Comment:
It dawns on me, now that we support both `:` and `;` as separators within `38;(2|5);.*` sequences, you have to take them into account in the `negate_pattern` list when trying to filter these kinds of sequences out. This seems like a potential footgun, could we document it here so that it's less likely people will forget this?